src/script/components/ZoomableImage/ZoomableImage.tsx

```diff
@@ -17,112 +17,230 @@
  *
  */
 
-import React, {HTMLProps, useState} from 'react';
+import React, {HTMLProps, RefObject, useEffect, useRef, useState} from 'react';
 
-import {imageStyle} from './ZoomableImage.style';
+import {containerStyle, imageStyle} from './ZoomableImage.style';
 
-function checkIfCanZoomImage(element: HTMLImageElement) {
-  if (element.naturalWidth <= element.width && element.naturalHeight <= element.height) {
-    return false;
+type Offset = {
+  x: number;
+  y: number;
+};
+
+const DEFAULT_OFFSET: Offset = {x: 0, y: 0};
+
+function calculateZoomRatio(element: HTMLImageElement) {
+  const parentElement = element.parentElement;
+
+  if (!parentElement) {
+    return 1;
+  }
+
+  const {offsetWidth: parentOffsetWidth, offsetHeight: parentOffsetHeight} = parentElement;
+  const {naturalWidth, naturalHeight} = element;
+
+  const widthRatio = parentOffsetWidth / naturalWidth;
+  const heightRatio = parentOffsetHeight / naturalHeight;
+  return Math.min(widthRatio, heightRatio);
+}
+
+// if we will add more image zooming, we need to pass 2 props, for check if is image is zoomed and imageScale
+function calculateMaxOffset(containerRef: RefObject<HTMLDivElement>, imgRef: RefObject<HTMLImageElement>) {
+  if (!containerRef.current || !imgRef.current) {
+    return {
+      maxXOffset: 0,
+      maxYOffset: 0,
+    };
   }
 
-  return element.naturalWidth !== element.offsetWidth || element.naturalHeight !== element.offsetHeight;
+  const containerRect = containerRef.current.getBoundingClientRect();
+
+  const maxXOffset = (containerRect.width - imgRef.current.naturalWidth) / 2;
+  const maxYOffset = (containerRect.height - imgRef.current.naturalHeight) / 2;
+
+  return {
+    maxXOffset: maxXOffset,
+    maxYOffset: maxYOffset,
+  };
 }
 
 type ZoomableImageProps = HTMLProps<HTMLImageElement>;
 
 export const ZoomableImage = (props: ZoomableImageProps) => {
+  const imageRef = useRef<HTMLImageElement | null>(null);
+  const containerRef = useRef<HTMLDivElement | null>(null);
+  const [imageRatio, setImageRatio] = useState(1);
+
+  const draggingRef = useRef(false);
+  const mouseDownRef = useRef(false);
+
   const [isZoomEnabled, setIsZoomEnabled] = useState<boolean>(false);
-  const [maxOffset, setMaxOffset] = useState({x: 0, y: 0});
 
-  const onButtonClick = (event: React.MouseEvent<HTMLImageElement>) => {
+  const [translateOffset, setTranslateOffset] = useState<Offset>(DEFAULT_OFFSET);
+  const [startOffset, setStartOffset] = useState<Offset>(DEFAULT_OFFSET);
+
+  const canZoomImage = imageRatio !== 1;
+  const zoomScale = isZoomEnabled ? 1 : imageRatio;
+
+  const handleMouseClick = (event: React.MouseEvent<HTMLDivElement>) => {
     const element = event.target as HTMLImageElement;
```

Comment:
Hmm, can we actually make sure that the image got clicked? 

Maybe something like:
```js
const element = event.target; // it's any event target here
const image = imageRef.current;

if(!image || element !== image){
  return;
}

// Now we're sure we've clicked on the image, we can type-safely operate on the ref value itself, eg.:
image.style.transition = '';

// [...]
```